tests/test_submain.py

```diff
@@ -1,47 +1,77 @@
 
-import validator.submain as submain
-from validator.errorbundler import ErrorBundle
+from validator import submain
 from validator.chromemanifest import ChromeManifest
-from validator.constants import *
-from helper import MockXPI
-
+from validator.errorbundler import ErrorBundle
+from .helper import MockXPI
 
-def test_prepare_package():
-    "Tests that the prepare_package function passes for valid data"
 
-    tp = submain.test_package
-    submain.test_package = lambda w, x, y, z, for_appversions: True
+@mock.patch("validator.submain.test_package")
+def test_prepare_package(test_package):
+    "Tests that the prepare_package function passes for valid data."
 
     err = ErrorBundle()
-    assert submain.prepare_package(err, "tests/resources/main/foo.xpi") == True
-    submain.test_package = tp
+    submain.prepare_package(err, "tests/resources/main/foo.xpi")
+    assert not err.failed()
 
 
-def test_validation_timeout():
-    tp = submain.test_inner_package
+@mock.patch("validator.submain.test_package")
+def test_validation_timeout(test_package):
     def slow(*args, **kw):
         time.sleep(1)
-    submain.test_inner_package = slow
+    test_package.side_effect = slow
+
+    err = ErrorBundle()
+    err.error(("an", "error"), "occurred")
+
+    submain.prepare_package(err, "tests/resources/main/foo.xpi", timeout=0.1)
+
+    # Make sure that our error got moved to the front of the list.
+    eq_(len(err.errors), 2)
+    eq_(err.errors[0]["id"],
+        ("validator", "unexpected_exception", "validation_timeout"))
+
+
+@mock.patch("validator.submain.test_package")
+@mock.patch("validator.errorbundler.log")
+def test_validation_error(log, test_package):
+    """Test that an unexpected exception during validation is turned into
+    an error message and logged."""
+
+    test_package.side_effect = Exception
+
     err = ErrorBundle()
-    submain.prepare_package(err, "tests/resources/main/foo.xpi",
-                            timeout=0.1)
-    submain.test_inner_package = tp
+    err.error(("an", "error"), "occurred")
 
-    assert len(err.errors) == 1
+    submain.prepare_package(err, "tests/resources/main/foo.xpi")
 
+    assert log.error.called
 
-def test_prepare_package_extension():
-    "Tests that bad extensions get outright rejections"
+    # Make sure that our error got moved to the front of the list.
+    eq_(len(err.errors), 2)
+    eq_(err.errors[0]["id"], ("validator", "unexpected_exception"))
 
-    assert submain.prepare_package(None, "foo/bar/test.foo") == False
 
-    ts = submain.test_search
-    submain.test_search = lambda x, y, z: True
-    assert submain.prepare_package(None, "foo/bar/test.xml") == True
-    submain.test_search = ts
+@mock.patch("validator.submain.test_search")
+def test_prepare_package_extension(test_search):
+    "Tests that bad extensions get outright rejections."
+
+    err = ErrorBundle()
+    submain.prepare_package(err, "foo/bar/test.foo")
+
+    assert not test_search.called
```

Comment:
A small comment explaining that it's not being called because it doesn't have a correct extension?